src/subprocess-posix.cc

```diff
@@ -343,9 +364,79 @@ void SubprocessSet::Clear() {
     // Since the foreground process is in our process group, it will receive
     // the interruption signal (i.e. SIGINT or SIGTERM) at the same time as us.
     if (!(*i)->use_console_)
-      kill(-(*i)->pid_, interrupted_);
+      Xkill(-(*i)->pid_, interrupted_);
   for (vector<Subprocess*>::iterator i = running_.begin();
        i != running_.end(); ++i)
     delete *i;
   running_.clear();
 }
+
+void SubprocessSet::Suspend() {
+  assert(IsSuspended());
+
+  // errno may change during the course of this function.
+  int saved_errno = errno;
+
+  // Suspend children.
+  for (vector<Subprocess*>::iterator i = running_.begin();
+       i != running_.end(); ++i)
+    if (!(*i)->use_console_)
+      Xkill(-(*i)->pid_, SIGTSTP);
+
+  // Tell users what we have done.
+  {
+    // We use write to bypass the buffer that may not be flush before
+    // we re-raise the SIGTSTP for us.
+    const char* msg = "\nninja: entire build suspended.\n";
+    // If the write fails, there's nothing we can do.
+    write(STDOUT_FILENO, msg, strlen(msg));
+  }
+
+  // Reset SIGTSTP handler to the default and re-raise the signal for us
+  // to trigger the default behavior when the signal will be unblock.
+  RestoreSigTSTPHandler();
+  raise(SIGTSTP);
+
+  // Unblock SIGTSP so that the pending signal we just raised can trigger
+  // the default handler of SIGTSTP we just restored. Thus, we get stopped
+  // right after sigprocmask call.
+  sigset_t tstp_mask;
+  sigset_t old_mask;
+  sigemptyset(&tstp_mask);
+  sigaddset(&tstp_mask, SIGTSTP);
+  if (sigprocmask(SIG_UNBLOCK, &tstp_mask, &old_mask) < 0)
+    Fatal("cannot unblock SIGTSTP: %s", strerror(errno));
+
+  // We are stopped here until we receive SIGCONT.
```

Comment:
Could you document how you know the program stops at exactly this line?

The documentation for sigprocmask http://man7.org/linux/man-pages/man2/sigprocmask.2.html doesn't give any indication that the program will process signals immediately upon unblocking them.

Reply:
From [signal(7)](http://man7.org/linux/man-pages/man7/signal.7.html)

```A signal may be blocked, which means that it will not be delivered until it is later unblocked.```

Apart from that and/or reading the source code I have no clue about when it is exactly delivered.

Reply:
Thanks for researching that.

Mind adding this information straight into the code?

Reply:
> The documentation for sigprocmask http://man7.org/linux/man-pages/man2/sigprocmask.2.html doesn't give any indication that the program will process signals immediately upon unblocking them.

That's an omission in the manual page then; see [POSIX](https://pubs.opengroup.org/onlinepubs/9799919799/functions/sigprocmask.html):
> If the argument _set_ is not a null pointer, after _pthread_sigmask()_ changes the currently blocked set of signals it shall determine whether there are any pending unblocked signals; if there are any, then at least one of those signals shall be delivered before the call to _pthread_sigmask()_ returns. [...] The _sigprocmask()_ function shall be equivalent to _pthread_sigmask()_, except that its behavior is unspecified if called from a multi-threaded process, and on error it returns `-1` and sets _errno_ to the error number instead of returning the error number directly.